docs/features/communication/abi_compatible_data_types/index.rst

```diff
@@ -0,0 +1,491 @@
+   :id: doc__abi_compatible_data_types
+   :status: valid
+   :safety: ASIL_B
+   :realizes: wp__feat_request
+   :tags: feature_request, change_management, communication, abi_compatible_data_types
+
+
+.. toctree::
+   :hidden:
+
+   requirements.rst
+
+
+Feature flag
+============
+
+To activate this feature, use the following feature flag:
+
+``experimental_abi_compatible_data_types``
+
+
+Abstract
+========
+
+This feature request defines a set of ABI-compatible data types and a runtime type description format to support zero-copy inter-process communication between C++17 and Rust 1.8x processes using the same endianness. It ensures consistent type layouts across languages by requiring fixed-size, statically allocated types without absolute pointers or language-specific metadata.
+
+The specification covers primitive types, structs, enums, arrays, and introduces ABI-stable representations for vectors, options, and results. A runtime-readable type description enables processes to interpret shared memory without compile-time access to type definitions.
+
+
+Motivation
+==========
+
+This feature request addresses specific challenges in achieving type compatibility within our inter-process communication (IPC) framework that leverages zero-copy shared memory mechanisms. Two essential scenarios are under evaluation:
+
+1. **ABI Compatibility**: Processes implemented in different programming languages (C++17 and Rust 1.8x) must interpret a shared memory location consistently as the same native type, provided both have compile-time access to the type definition. This scenario eliminates serialization overhead and allows direct memory access.
+
+2. **Type Description**: Processes responsible for translating data between internal ABI-compatible formats and external serialization formats should perform this translation without compile-time knowledge of type definitions. This runtime interpretation capability is critical for scalability, ensuring that gateway processes and tools for recording/replaying/inspecting data can dynamically handle new data types without recompilation.
+
+
+ABI Compatibility
+-----------------
+
+Our communication feature relies on shared memory to transfer data between processes. For effective zero-copy data exchange, processes written in C++17 and Rust 1.8x must inherently understand the data at shared memory locations identically. Achieving this requires ensuring that data types have consistent, fixed-size memory layouts.
+
+This evaluation initially targets the following process configurations:
+
+* Processes running on the same operating system.
+* Processes running on different operating systems but under the same hypervisor.
+
+Supporting different endianness between processes is explicitly out of scope, as it inherently demands bit manipulation, effectively requiring serialization.
+Different bit widths, however, are implicitly supported by specifying the width of all types and excluding word-size integers.
```

Comment:
How would this play with alignments? I would imagine we end up in quite a huge mess if we have two parties accessing the same data but with a different bit width.

Maybe you mean that we want to support that our ABI compatible datatypes work on different bit widths, as long as all participants have the same bit width. If so, we should rephrase to make this more explicit.